File: 2016/25.py
import math

with open("input_25.txt") as f:
    lines = f.readlines()

for i in range(len(lines)):
    lines[i] = lines[i].split("\n")[0]

a = 0
while True:
    registry = {"a":a, "b":0, "c":0, "d":0}

    outs = []
    line = 0
    n = 0
    while line < len(lines):
        parts = lines[line].split()
        cmd = parts[0]
        params = parts[1:]
        jump = False
        
        if cmd == "cpy":
            try:
                x = int(params[0])
            except ValueError:
                x = registry[params[0]]
            registry[params[1]] = x
        elif cmd == "inc":
            registry[params[0]] += 1
        elif cmd == "dec":
            registry[params[0]] -= 1
        elif cmd == "out":
            try:
                x = int(params[0])        
            except ValueError:
                x = registry[params[0]]
            outs.append(x)
            n += 1
            if n == 10:
                #print (outs)
                if outs == [0,1,0,1,0,1,0,1,0,1]:
                    print("🎁🎄Last part 2016: {}".format(a))
                    import sys
                    sys.exit()
                break
        elif cmd == "jnz":
            try:
                x = int(params[0])        
            except ValueError:
                x = registry[params[0]]
                
            try:
                y = int(params[1])        
            except ValueError:
                y = registry[params[1]]
            if x != 0:
                jump = True
                line += int(y)
        elif cmd == "sum":
            try:
                y = int(params[1])        
            except ValueError:
                y = registry[params[1]]
                
            registry[params[0]] += y

        elif cmd == "fac":
            try:
                y = int(params[1])        
            except ValueError:
                y = registry[params[1]]

            registry[params[0]] = math.factorial(y)
        elif cmd == "mlt":
            try:
                x = int(params[0])        
            except ValueError:
                x = registry[params[0]]

            try:
                y = int(params[1])        
            except ValueError:
                y = registry[params[1]]

            try:
                z = int(params[2])        
            except ValueError:
                z = registry[params[2]]
            registry[params[0]] = y*z
            
        elif cmd == "tgl":
            try:
                x = int(params[0])        
            except ValueError:
                x = registry[params[0]]
                
            if 0 <= line + x < len(lines):
                l = lines[line + x]
                nparams = len(l.split())
                if nparams == 2:
                    if l.startswith("inc"):
                        lines[line+x] = l.replace("inc", "dec")
                    else:
                        lines[line+x] = "inc" + l[3:]
                elif nparams == 3:
                    if l.startswith("jnz"):
                        lines[line+x] = l.replace("jnz", "cpy")
                    else:
                        lines[line+x] = "jnz" + l[3:]
                        
        if not jump:
            line += 1
    a += 1

File: 2019/25.jl
include("utils.jl")
using .AoCUtils

register = Dict("a"=>0, "b"=>0, "c"=>0, "d"=>0)

function check_str2(a)
    return tryparse(Float64, a) !== nothing
end

function parseArg(arg)
    if check_str2(arg)
        return parse(Int, arg)
    else
        return register[arg]
    end
end

function compiler(lines)
    stdout = []
    N = 0
    cline = 1
    while true
        l = lines[cline]

        if startswith(l, "cpy")
            args = split(l, " ")[2:end]
            register[args[2]] = parseArg(args[1])
            cline += 1
        elseif startswith(l, "inc")
            arg = split(l, " ")[end]
            register[arg] += 1
            cline += 1
        elseif startswith(l, "dec")
            arg = split(l, " ")[end]
            register[arg] -= 1
            cline += 1
        elseif startswith(l, "jnz")
            args = split(l, " ")[2:end]
            if parseArg(args[1]) != 0
                cline += parseArg(args[2])
            else
                cline += 1
            end
        elseif startswith(l, "sum")
            args = split(l, " ")[2:end]
            register[args[1]] += parseArg(args[2])
            cline += 1
        elseif startswith(l, "fac")
            args = split(l, " ")[2:end]
            register[args[1]] = factorial(parseArg(args[2]))
            cline += 1
        elseif startswith(l, "mlt")
            args = split(l, " ")[2:end]
            register[args[1]] = parseArg(args[2]) * parseArg(args[3])
            cline += 1
        elseif startswith(l, "out")
            N += 1
            args = split(l, " ")[2:end]
            push!(stdout, parseArg(args[1]))
            if N == 10
                if all(stdout .== [0,1,0,1,0,1,0,1,0,1])
                    return true
                else
                    return false
                end
            end
            cline += 1    
        elseif startswith(l, "tgl")
            arg = parseArg(split(l, " ")[2])
            if  1 <= cline + arg <= length(lines)
                tgl_line = lines[cline + arg]
                if startswith(tgl_line, "tgl")
                    lines[cline + arg] = replace(tgl_line, "tgl"=>"inc")
                elseif startswith(tgl_line, "inc")
                    lines[cline + arg] = replace(tgl_line, "inc"=>"dec")
                elseif startswith(tgl_line, "dec")
                    lines[cline + arg] = replace(tgl_line, "dec"=>"inc")
                elseif startswith(tgl_line, "cpy")
                    lines[cline + arg] = replace(tgl_line, "cpy"=>"jnz")
                elseif startswith(tgl_line, "jnz")
                    lines[cline + arg] = replace(tgl_line, "jnz"=>"cpy")
                end
            end
            cline += 1
        end

        if cline < 1 || cline >length(lines)
            break
        end
    end

    println("🎅 $(register["a"])")
end

lines = AoCUtils.readLines("input_25.txt")
for i in 1:500
    register["a"] = i
    if compiler(lines)
        println("🎅🛷 $(i)")
        break
    end 
end
